### ISSUE TITLE
Unreachable `isinstance` by subclassing

### ISSUE THREAD

**Issue by sylee957** [microsoft/pyright#5764]
**Describe the bug**

**Code or Screenshots**

```python
from typing import Any


class Animal:
    pass


class Dog(Animal):
    pass


def test_1(x: Animal | Dog):
    if isinstance(x, Animal):
        reveal_type(x)
        return "Animal"
    if isinstance(x, Dog):
        reveal_type(x)
        return "Bark"


def test_2(x: Any):
    if isinstance(x, Animal):
        reveal_type(x)
        return "Animal"
    if isinstance(x, Dog):
        reveal_type(x)
        return "Bark"
```

`test_1` gives the unreachability error for second isinstance check, however, type checking `test_2` can't infer that the `isinstance(x, Dog)` is unreachable.
This is a code example from some incorrectly implemented single dispatching, which is not easy to discover the mistake like this. 
I'm not sure if this is a limitation about `pyright` or if something is special about `Any` that makes the check weaker.

**VS Code extension or command-line**

I'm using pyright `1.1.323` with npm.

**Comment by erictraut** [comment]
This is working as designed, so I don't consider this a bug.

Pyright doesn't implement any "reachability checks". I presume you're referring to the `reportUnnecessaryIsinstance` check. This check is based on whether the `isinstance` call has any effect on type narrowing.

In the first example, pyright has knowledge of the class hierarchy, so it can determine that `isinstance(x, Animal)` will necessarily be true for all values described by the annotation `Animal | Dog`. That's why it's flagged by the `reportUnnecessaryIsinstance` check, which is unnecessary in this case.

In the second example, the `isinstance(x, Animal)` can narrow `x` to `Animal` in the positive case, but it can't be narrowed in the negative ("else") case, so it remains `Any`. That means the second `isinstance` check is still applicable (it's not unnecessary).

**Comment by sylee957** [comment]
I see, I think that I had found some evidence why the design with dispatching with subclassing can be less favored than disjoint unions
